### PR TITLE
WIP Unix domain socket support

### PR DESCRIPTION
**Pull request by mattrobenolt**
Addressing #1465

From my perspective, this seems pretty feature complete. We've also been using our version of this in production for a bit now without any issues. (https://github.com/getsentry/sentry/blob/master/src/sentry/net/http.py#L156)

I've made a couple design decisions based on the discussion in the other ticket:

* Added a `UnixHTTPPoolManager` which provides a `http+unix` scheme and a URI quoted file path so a URI like `http+unix://%2Ftmp%2Fserver.sock/hello` works. This also subclasses normal `PoolManager` so technically this one could just be swapped in to get full http/http/http+unix support.

* Two different `ConnectionPool` classes. One takes the URI quoted version of the path (`UnixHostHTTPConnectionPool`) and one that takes `socket_path` directly (`UnixHTTPConnectionPool`). The first one is to support the API needed for the `PoolManager` and is simply responsible for unquoting the path and passing it along.

I have started modifying the test suite to accommodate a Unix domain socket dummyserver, but haven't added any real tests yet.

I'm ready for feedback on APIs here and implementation, and will spend time figuring out tests after we are comfortable. I've tested stuff manually so far, so it's working in current state.

Thanks everyone! ❤️

### CODE REVIEW

**Comment by pquentin** [comment]
@mattrobenolt There's a flake8 failure left, and I guess you need to make sure Windows does not see this feature!

**Comment by mattrobenolt** [comment]
Yeah, I also don’t have tests. Hence the WIP. Was hoping to get early feedback before investing more time and people disagreeing with the direction. If all is good, I’m happy to polish things up.

**Comment by mattrobenolt** [comment]
For sure, just wanted to actually get the code out there since the direction was theoretical without an implementation. Now I have an implementation.

**Comment by theacodes** [comment]
The general idea and approach is good with me. :)

**Comment by SethMichaelLarson** [comment]
The implementation looks good to me too, just waiting for some test cases. :) And I know this is a module we can probably cover 100% so maybe we need to be more granular when it comes to exempting things within `urllib3/contrib`?

**Comment by mattrobenolt** [comment]
Cool, I'll get things wrapped up here.

With regards to Windows, obviously we can skip tests for Windows since it won't work, but do we have opinions about just letting the entire class just explode if someone tries to use it on Windows? Or should we be nicer and provide a stub class that says something like, "Unix sockets don't work on Windows". Or just let the module itself ImportError if trying to import on Windows.

I also don't have access to a Windows machine, so will be flying a bit blind there.

**Comment by SethMichaelLarson** [comment]
Going from our `urllib3.contrib.socks` it looks like we give a warning if we can't import things we need (in this case `AF_UNIX`)? Is this path something we want to continue?

**Comment by sethmlarson** [comment]
@mattrobenolt did you still want to land this patch?

File: dummyserver/server.py [hunk 1]
```diff
@@ -267,12 +272,16 @@ def get_unreachable_address():
 if __name__ == '__main__':
     # For debugging dummyserver itself - python -m dummyserver.server
     from .testcase import TestingApp
-    host = '127.0.0.1'
+    # host = '127.0.0.1'
```
Comment (sethmlarson) [inline dummyserver/server.py]: Let's revert this change.

File: dummyserver/server.py [hunk 2]
```diff
@@ -237,8 +238,12 @@ def run_tornado_app(app, io_loop, certs, scheme, host):
     else:
         http_server = tornado.httpserver.HTTPServer(app)
 
-    sockets = bind_sockets(None, address=host)
-    port = sockets[0].getsockname()[1]
+    if host[:1] == '/':
```
Comment (sethmlarson) [inline dummyserver/server.py]: [:1] is the same as [0] so let's use the latter for better readability

File: dummyserver/server.py [hunk 3]
```diff
@@ -267,12 +272,16 @@ def get_unreachable_address():
...
     app = tornado.web.Application([(r".*", TestingApp)])
     server, port = run_tornado_app(app, io_loop, None,
                                    'http', host)
     server_thread = run_loop_in_thread(io_loop)
 
-    print("Listening on http://{host}:{port}".format(host=host, port=port))
+    if port:
+        print("Listening on http://{host}:{port}".format(host=host, port=port))
+    else:
+        print("Listening on unix@{host}".format(host=host))
```
Comment (sethmlarson) [inline dummyserver/server.py]: Let's change this scheme for this message to be the same as what we're using as our scheme below (ie `http+unix`)

File: test/with_dummyserver/test_unix.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    @classmethod
+    def tearDownClass(cls):
+        cls._stop_server()
+        try:
+            os.remove(cls.host)
+        except OSError:
+            pass
+
+
+class TestUnixHTTPConnectionPool(UnixHTTPDummyServerTestCase):
```
Comment (sethmlarson) [inline test/with_dummyserver/test_unix.py]: Can we somehow reuse the tests already written for our HTTPConnectionPool here? Would be great to have them generic enough to reuse.

File: dummyserver/server.py (hunk 2 above)
Comment (sethmlarson) [reply dummyserver/server.py]: I agree they don't have the same behavior in general, but isn't the type of the `host` parameter in Python 3 a `str` not `bytes` here and empty strings invalid hosts anyways?

**Comment by sethmlarson** [comment]
Take your time, thanks for tackling this for us!

File: src/urllib3/contrib/unix.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    """
+    A thread-safe connection pool for one Unix socket, addressed by URI.
+    :param host:
+        URI quoted path to be used for this HTTP Connection (e.g. "%2Fvar%2Frun%2Fserver.sock").
+    """
+    scheme = 'http+unix'
+
+    # port exists for API compatibility, but is ignored
+    def __init__(self, host, port=None, **conn_kw):
+        socket_path = unquote_plus(host)
```
Comment (sethmlarson) [inline src/urllib3/contrib/unix.py]: I don't think `unquote_plus` is what we want here, `+` is allowed in the authority section of URLs and we don't want to decode a `+` to a space character.

**Comment by pquentin** [comment]
Closing as this pull request is now stale (and needs changes).

@mattrobenolt Feel free to reopen if you resume this work.

If someone else is interested, please take this work and open a new pull request. Thanks!
